FATAL: class jenkins.plugins.slack.SlackNotifier$SlackJobProperty is missing its descriptor since 2.0

Since the upgrade to 2.0 we have the following error for all our builds:

```
FATAL: class jenkins.plugins.slack.SlackNotifier$SlackJobProperty is missing its descriptor
java.lang.AssertionError: class jenkins.plugins.slack.SlackNotifier$SlackJobProperty is missing its descriptor
    at jenkins.model.Jenkins.getDescriptorOrDie(Jenkins.java:1223)
    at hudson.model.JobProperty.getDescriptor(JobProperty.java:105)
    at hudson.model.JobProperty.getDescriptor(JobProperty.java:79)
    at hudson.model.Descriptor.toMap(Descriptor.java:973)
    at hudson.model.Job.getProperties(Job.java:558)
    at hudson.model.Build$BuildExecution.post2(Build.java:187)
    at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:668)
    at hudson.model.Run.execute(Run.java:1763)
    at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
    at hudson.model.ResourceController.execute(ResourceController.java:98)
    at hudson.model.Executor.run(Executor.java:410)
```

Downgrade to 1.8.1 seems to solve the issue.
Jenkins version: 1.652
# See possible fix

edit by @samrocketman: See #191 for possible fixes for this issue.

@OwensCode, this is in response to your request on #193. It seemed more appropriate to keep the discussion here.

I am sorry, but I do not have my configuration from prior to fixing it as described. 

Here is a snippet from the log file, which suggests that the migration process was started, and is quickly followed by the failed attempt to save the config.

I'd be happy to send you the full log file directly, but I am concerned about posting it publically as it looks like it contains sensitive info, such as keys. 

```
Mar 16, 2016 7:06:07 PM jenkins.plugins.slack.SlackNotifier$Migrator onLoaded
INFO: Starting Settings Migration Process
Mar 16, 2016 7:06:07 PM jenkins.plugins.slack.SlackNotifier$Migrator onLoaded
INFO: processing Job: feature%2Fgrade-column-mapping
Mar 16, 2016 7:06:07 PM jenkins.plugins.slack.SlackNotifier$Migrator onLoaded
INFO: Configuration updated successfully
Mar 16, 2016 7:06:07 PM jenkins.plugins.slack.SlackNotifier$Migrator onLoaded
INFO: processing Job: feature%2Frestructure-build-gradle
Mar 16, 2016 7:06:07 PM jenkins.plugins.slack.SlackNotifier$Migrator onLoaded
INFO: Configuration updated successfully
Mar 16, 2016 7:06:07 PM jenkins.plugins.slack.SlackNotifier$Migrator onLoaded
INFO: processing Job: master
Mar 16, 2016 7:06:07 PM jenkins.plugins.slack.SlackNotifier$Migrator onLoaded
INFO: Configuration updated successfully
Mar 16, 2016 7:06:07 PM jenkins.InitReactorRunner$1 onAttained
INFO: Started initialization
Mar 16, 2016 7:06:07 PM jenkins.InitReactorRunner$1 onAttained
INFO: Listed all plugins
Mar 16, 2016 7:06:07 PM jenkins.InitReactorRunner$1 onAttained
INFO: Prepared all plugins
Mar 16, 2016 7:06:07 PM jenkins.InitReactorRunner$1 onAttained
INFO: Started all plugins
Mar 16, 2016 7:06:07 PM jenkins.InitReactorRunner$1 onAttained
INFO: Augmented all extensions
Mar 16, 2016 7:06:07 PM jenkins.InitReactorRunner$1 onAttained
INFO: Loaded all jobs
Mar 16, 2016 7:06:07 PM jenkins.InitReactorRunner$1 onAttained
INFO: Completed initialization
Mar 16, 2016 7:06:07 PM hudson.WebAppMain$3 run
INFO: Jenkins is fully up and running
Mar 16, 2016 7:08:13 PM hudson.ExpressionFactory2$JexlExpression evaluate
WARNING: Caught exception evaluating: it.properties.values() in /job/Darkwing/branch/feature%252Frestructure-build-gradle/. Reason: java.lang.reflect.InvocationTargetException
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.commons.jexl.util.PropertyExecutor.execute(PropertyExecutor.java:125)
	at org.apache.commons.jexl.util.introspection.UberspectImpl$VelGetterImpl.invoke(UberspectImpl.java:314)
	at org.apache.commons.jexl.parser.ASTArrayAccess.evaluateExpr(ASTArrayAccess.java:185)
	at org.apache.commons.jexl.parser.ASTIdentifier.execute(ASTIdentifier.java:75)
	at org.apache.commons.jexl.parser.ASTReference.execute(ASTReference.java:83)
	at org.apache.commons.jexl.parser.ASTReference.value(ASTReference.java:57)
	at org.apache.commons.jexl.parser.ASTReferenceExpression.value(ASTReferenceExpression.java:51)
	at org.apache.commons.jexl.ExpressionImpl.evaluate(ExpressionImpl.java:80)
	at hudson.ExpressionFactory2$JexlExpression.evaluate(ExpressionFactory2.java:74)
	at org.apache.commons.jelly.expression.ExpressionSupport.evaluateRecurse(ExpressionSupport.java:61)
	at org.apache.commons.jelly.expression.ExpressionSupport.evaluateAsIterator(ExpressionSupport.java:94)
	at org.apache.commons.jelly.tags.core.ForEachTag.doTag(ForEachTag.java:89)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.JellyViewScript.run(JellyViewScript.java:95)
	at org.kohsuke.stapler.jelly.IncludeTag.doTag(IncludeTag.java:147)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.CallTagLibScript$1.run(CallTagLibScript.java:99)
	at org.apache.commons.jelly.tags.define.InvokeBodyTag.doTag(InvokeBodyTag.java:91)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$1.run(CoreTagLibrary.java:98)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.CallTagLibScript.run(CallTagLibScript.java:120)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.CallTagLibScript$1.run(CallTagLibScript.java:99)
	at org.apache.commons.jelly.tags.define.InvokeBodyTag.doTag(InvokeBodyTag.java:91)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.CallTagLibScript.run(CallTagLibScript.java:120)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.JellyViewScript.run(JellyViewScript.java:95)
	at org.kohsuke.stapler.jelly.DefaultScriptInvoker.invokeScript(DefaultScriptInvoker.java:63)
	at org.kohsuke.stapler.jelly.DefaultScriptInvoker.invokeScript(DefaultScriptInvoker.java:53)
	at org.kohsuke.stapler.jelly.JellyClassTearOff.serveIndexJelly(JellyClassTearOff.java:112)
	at org.kohsuke.stapler.jelly.JellyFacet.handleIndexRequest(JellyFacet.java:127)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:735)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.MetaClass$5.doDispatch(MetaClass.java:233)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:746)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.MetaClass$5.doDispatch(MetaClass.java:233)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:746)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:649)
	at org.kohsuke.stapler.Stapler.service(Stapler.java:238)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:848)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:686)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1494)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:134)
	at hudson.util.PluginServletFilter.doFilter(PluginServletFilter.java:125)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at hudson.security.csrf.CrumbFilter.doFilter(CrumbFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:84)
	at hudson.security.UnwrapSecurityExceptionFilter.doFilter(UnwrapSecurityExceptionFilter.java:51)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:117)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.providers.anonymous.AnonymousProcessingFilter.doFilter(AnonymousProcessingFilter.java:125)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.rememberme.RememberMeProcessingFilter.doFilter(RememberMeProcessingFilter.java:142)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.AbstractProcessingFilter.doFilter(AbstractProcessingFilter.java:271)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.BasicHeaderProcessor.doFilter(BasicHeaderProcessor.java:93)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.context.HttpSessionContextIntegrationFilter.doFilter(HttpSessionContextIntegrationFilter.java:249)
	at hudson.security.HttpSessionContextIntegrationFilter2.doFilter(HttpSessionContextIntegrationFilter2.java:67)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:76)
	at hudson.security.HudsonFilter.doFilter(HudsonFilter.java:171)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at org.kohsuke.stapler.compression.CompressionFilter.doFilter(CompressionFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:81)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter(DiagnosticThreadNameFilter.java:30)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1474)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:499)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:137)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:533)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:231)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1086)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:428)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:193)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1020)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:135)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:116)
	at org.eclipse.jetty.server.Server.handle(Server.java:370)
	at org.eclipse.jetty.server.AbstractHttpConnection.handleRequest(AbstractHttpConnection.java:489)
	at org.eclipse.jetty.server.AbstractHttpConnection.headerComplete(AbstractHttpConnection.java:949)
	at org.eclipse.jetty.server.AbstractHttpConnection$RequestHandler.headerComplete(AbstractHttpConnection.java:1011)
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:644)
	at org.eclipse.jetty.http.HttpParser.parseAvailable(HttpParser.java:235)
	at org.eclipse.jetty.server.AsyncHttpConnection.handle(AsyncHttpConnection.java:82)
	at org.eclipse.jetty.io.nio.SelectChannelEndPoint.handle(SelectChannelEndPoint.java:668)
	at org.eclipse.jetty.io.nio.SelectChannelEndPoint$1.run(SelectChannelEndPoint.java:52)
	at winstone.BoundedExecutorService$1.run(BoundedExecutorService.java:77)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.AssertionError: class jenkins.plugins.slack.SlackNotifier$SlackJobProperty is missing its descriptor
	at jenkins.model.Jenkins.getDescriptorOrDie(Jenkins.java:1257)
	at hudson.model.JobProperty.getDescriptor(JobProperty.java:105)
	at hudson.model.JobProperty.getDescriptor(JobProperty.java:79)
	at hudson.model.Descriptor.toMap(Descriptor.java:973)
	at hudson.model.Job.getProperties(Job.java:558)
	... 119 more

Mar 16, 2016 7:17:32 PM hudson.ExpressionFactory2$JexlExpression evaluate
WARNING: Caught exception evaluating: attrs.instances ?: instance[field] ?: descriptor['default'+h.capitalize(field)] in /job/Darkwing/configure. Reason: java.lang.reflect.InvocationTargetException
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.commons.jexl.util.PropertyExecutor.execute(PropertyExecutor.java:125)
	at org.apache.commons.jexl.util.introspection.UberspectImpl$VelGetterImpl.invoke(UberspectImpl.java:314)
	at org.apache.commons.jexl.parser.ASTArrayAccess.evaluateExpr(ASTArrayAccess.java:185)
	at org.apache.commons.jexl.parser.ASTArrayAccess.value(ASTArrayAccess.java:118)
	at org.apache.commons.jexl.parser.ASTReference.execute(ASTReference.java:77)
	at org.apache.commons.jexl.parser.ASTReference.value(ASTReference.java:57)
	at org.apache.commons.jexl.parser.ASTElvisNode.value(ASTElvisNode.java:37)
	at org.apache.commons.jexl.parser.ASTExpression.value(ASTExpression.java:54)
	at org.apache.commons.jexl.parser.ASTElvisNode.value(ASTElvisNode.java:40)
	at org.apache.commons.jexl.parser.ASTExpression.value(ASTExpression.java:54)
	at org.apache.commons.jexl.parser.ASTExpressionExpression.value(ASTExpressionExpression.java:56)
	at org.apache.commons.jexl.ExpressionImpl.evaluate(ExpressionImpl.java:80)
	at hudson.ExpressionFactory2$JexlExpression.evaluate(ExpressionFactory2.java:74)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$3.run(CoreTagLibrary.java:134)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.CallTagLibScript.run(CallTagLibScript.java:120)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.CallTagLibScript$1.run(CallTagLibScript.java:99)
	at org.apache.commons.jelly.tags.define.InvokeBodyTag.doTag(InvokeBodyTag.java:91)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.CallTagLibScript$1.run(CallTagLibScript.java:99)
	at org.apache.commons.jelly.tags.define.InvokeBodyTag.doTag(InvokeBodyTag.java:91)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.TagSupport.invokeBody(TagSupport.java:161)
	at org.apache.commons.jelly.tags.core.WhenTag.doTag(WhenTag.java:46)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.TagSupport.invokeBody(TagSupport.java:161)
	at org.apache.commons.jelly.tags.core.ChooseTag.doTag(ChooseTag.java:38)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.CallTagLibScript.run(CallTagLibScript.java:120)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.CallTagLibScript.run(CallTagLibScript.java:120)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.JellyViewScript.run(JellyViewScript.java:95)
	at org.kohsuke.stapler.jelly.IncludeTag.doTag(IncludeTag.java:147)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.CallTagLibScript$1.run(CallTagLibScript.java:99)
	at org.apache.commons.jelly.tags.define.InvokeBodyTag.doTag(InvokeBodyTag.java:91)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.CallTagLibScript.run(CallTagLibScript.java:120)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.JellyViewScript.run(JellyViewScript.java:95)
	at org.kohsuke.stapler.jelly.IncludeTag.doTag(IncludeTag.java:147)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.JellyViewScript.run(JellyViewScript.java:95)
	at org.kohsuke.stapler.jelly.IncludeTag.doTag(IncludeTag.java:147)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.CallTagLibScript$1.run(CallTagLibScript.java:99)
	at org.apache.commons.jelly.tags.define.InvokeBodyTag.doTag(InvokeBodyTag.java:91)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.CallTagLibScript.run(CallTagLibScript.java:120)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.CallTagLibScript$1.run(CallTagLibScript.java:99)
	at org.apache.commons.jelly.tags.define.InvokeBodyTag.doTag(InvokeBodyTag.java:91)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$1.run(CoreTagLibrary.java:98)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.CallTagLibScript.run(CallTagLibScript.java:120)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.CallTagLibScript$1.run(CallTagLibScript.java:99)
	at org.apache.commons.jelly.tags.define.InvokeBodyTag.doTag(InvokeBodyTag.java:91)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.CallTagLibScript.run(CallTagLibScript.java:120)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.JellyViewScript.run(JellyViewScript.java:95)
	at org.kohsuke.stapler.jelly.DefaultScriptInvoker.invokeScript(DefaultScriptInvoker.java:63)
	at org.kohsuke.stapler.jelly.DefaultScriptInvoker.invokeScript(DefaultScriptInvoker.java:53)
	at org.kohsuke.stapler.jelly.JellyFacet$1.dispatch(JellyFacet.java:95)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:746)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.MetaClass$5.doDispatch(MetaClass.java:233)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:746)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:649)
	at org.kohsuke.stapler.Stapler.service(Stapler.java:238)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:848)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:686)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1494)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:134)
	at hudson.util.PluginServletFilter.doFilter(PluginServletFilter.java:125)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at hudson.security.csrf.CrumbFilter.doFilter(CrumbFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:84)
	at hudson.security.UnwrapSecurityExceptionFilter.doFilter(UnwrapSecurityExceptionFilter.java:51)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:117)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.providers.anonymous.AnonymousProcessingFilter.doFilter(AnonymousProcessingFilter.java:125)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.rememberme.RememberMeProcessingFilter.doFilter(RememberMeProcessingFilter.java:142)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.AbstractProcessingFilter.doFilter(AbstractProcessingFilter.java:271)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.BasicHeaderProcessor.doFilter(BasicHeaderProcessor.java:93)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.context.HttpSessionContextIntegrationFilter.doFilter(HttpSessionContextIntegrationFilter.java:249)
	at hudson.security.HttpSessionContextIntegrationFilter2.doFilter(HttpSessionContextIntegrationFilter2.java:67)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:76)
	at hudson.security.HudsonFilter.doFilter(HudsonFilter.java:171)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at org.kohsuke.stapler.compression.CompressionFilter.doFilter(CompressionFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:81)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter(DiagnosticThreadNameFilter.java:30)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1474)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:499)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:137)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:533)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:231)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1086)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:428)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:193)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1020)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:135)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:116)
	at org.eclipse.jetty.server.Server.handle(Server.java:370)
	at org.eclipse.jetty.server.AbstractHttpConnection.handleRequest(AbstractHttpConnection.java:489)
	at org.eclipse.jetty.server.AbstractHttpConnection.headerComplete(AbstractHttpConnection.java:949)
	at org.eclipse.jetty.server.AbstractHttpConnection$RequestHandler.headerComplete(AbstractHttpConnection.java:1011)
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:644)
	at org.eclipse.jetty.http.HttpParser.parseAvailable(HttpParser.java:235)
	at org.eclipse.jetty.server.AsyncHttpConnection.handle(AsyncHttpConnection.java:82)
	at org.eclipse.jetty.io.nio.SelectChannelEndPoint.handle(SelectChannelEndPoint.java:668)
	at org.eclipse.jetty.io.nio.SelectChannelEndPoint$1.run(SelectChannelEndPoint.java:52)
	at winstone.BoundedExecutorService$1.run(BoundedExecutorService.java:77)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.AssertionError: class jenkins.plugins.slack.SlackNotifier$SlackJobProperty is missing its descriptor
	at jenkins.model.Jenkins.getDescriptorOrDie(Jenkins.java:1257)
	at hudson.model.JobProperty.getDescriptor(JobProperty.java:105)
	at hudson.model.JobProperty.getDescriptor(JobProperty.java:79)
	at hudson.model.Descriptor.toMap(Descriptor.java:973)
	at hudson.model.Job.getProperties(Job.java:558)
	... 165 more

Mar 16, 2016 7:17:58 PM org.eclipse.jetty.util.log.JavaUtilLog warn
WARNING: Error while serving https://jenkins.alltheducks.com/job/Darkwing/configSubmit
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.kohsuke.stapler.Function$InstanceFunction.invoke(Function.java:324)
	at org.kohsuke.stapler.interceptor.RequirePOST$Processor.invoke(RequirePOST.java:52)
	at org.kohsuke.stapler.PreInvokeInterceptedFunction.invoke(PreInvokeInterceptedFunction.java:26)
	at org.kohsuke.stapler.Function.bindAndInvoke(Function.java:167)
	at org.kohsuke.stapler.Function.bindAndInvokeAndServeResponse(Function.java:100)
	at org.kohsuke.stapler.MetaClass$1.doDispatch(MetaClass.java:124)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:746)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.MetaClass$5.doDispatch(MetaClass.java:233)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:746)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:649)
	at org.kohsuke.stapler.Stapler.service(Stapler.java:238)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:848)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:686)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1494)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:134)
	at hudson.util.PluginServletFilter.doFilter(PluginServletFilter.java:125)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at hudson.security.csrf.CrumbFilter.doFilter(CrumbFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:84)
	at hudson.security.UnwrapSecurityExceptionFilter.doFilter(UnwrapSecurityExceptionFilter.java:51)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:117)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.providers.anonymous.AnonymousProcessingFilter.doFilter(AnonymousProcessingFilter.java:125)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.rememberme.RememberMeProcessingFilter.doFilter(RememberMeProcessingFilter.java:142)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.AbstractProcessingFilter.doFilter(AbstractProcessingFilter.java:271)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.BasicHeaderProcessor.doFilter(BasicHeaderProcessor.java:93)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.context.HttpSessionContextIntegrationFilter.doFilter(HttpSessionContextIntegrationFilter.java:249)
	at hudson.security.HttpSessionContextIntegrationFilter2.doFilter(HttpSessionContextIntegrationFilter2.java:67)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:76)
	at hudson.security.HudsonFilter.doFilter(HudsonFilter.java:171)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at org.kohsuke.stapler.compression.CompressionFilter.doFilter(CompressionFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:81)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter(DiagnosticThreadNameFilter.java:30)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1474)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:499)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:137)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:533)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:231)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1086)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:428)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:193)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1020)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:135)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:116)
	at org.eclipse.jetty.server.Server.handle(Server.java:370)
	at org.eclipse.jetty.server.AbstractHttpConnection.handleRequest(AbstractHttpConnection.java:489)
	at org.eclipse.jetty.server.AbstractHttpConnection.content(AbstractHttpConnection.java:960)
	at org.eclipse.jetty.server.AbstractHttpConnection$RequestHandler.content(AbstractHttpConnection.java:1021)
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:865)
	at org.eclipse.jetty.http.HttpParser.parseAvailable(HttpParser.java:235)
	at org.eclipse.jetty.server.AsyncHttpConnection.handle(AsyncHttpConnection.java:82)
	at org.eclipse.jetty.io.nio.SelectChannelEndPoint.handle(SelectChannelEndPoint.java:668)
	at org.eclipse.jetty.io.nio.SelectChannelEndPoint$1.run(SelectChannelEndPoint.java:52)
	at winstone.BoundedExecutorService$1.run(BoundedExecutorService.java:77)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.AssertionError: class jenkins.plugins.slack.SlackNotifier$SlackJobProperty is missing its descriptor
	at jenkins.model.Jenkins.getDescriptorOrDie(Jenkins.java:1257)
	at hudson.model.JobProperty.getDescriptor(JobProperty.java:105)
	at hudson.model.JobProperty.getDescriptor(JobProperty.java:79)
	at hudson.util.DescribableList.get(DescribableList.java:127)
	at hudson.util.DescribableList.rebuild(DescribableList.java:168)
	at hudson.model.Job.doConfigSubmit(Job.java:1218)
	at hudson.model.AbstractProject.doConfigSubmit(AbstractProject.java:795)
	at com.github.mjdetullio.jenkins.plugins.multibranch.AbstractMultiBranchProject.submit(AbstractMultiBranchProject.java:538)
	at com.cloudbees.hudson.plugins.folder.AbstractFolder.doConfigSubmit(AbstractFolder.java:689)
	at com.cloudbees.hudson.plugins.folder.computed.ComputedFolder.doConfigSubmit(ComputedFolder.java:216)
	... 75 more

Mar 16, 2016 7:18:34 PM hudson.ExpressionFactory2$JexlExpression evaluate
WARNING: Caught exception evaluating: attrs.instances ?: instance[field] ?: descriptor['default'+h.capitalize(field)] in /job/Darkwing/configure. Reason: java.lang.reflect.InvocationTargetException
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.commons.jexl.util.PropertyExecutor.execute(PropertyExecutor.java:125)
	at org.apache.commons.jexl.util.introspection.UberspectImpl$VelGetterImpl.invoke(UberspectImpl.java:314)
	at org.apache.commons.jexl.parser.ASTArrayAccess.evaluateExpr(ASTArrayAccess.java:185)
	at org.apache.commons.jexl.parser.ASTArrayAccess.value(ASTArrayAccess.java:118)
	at org.apache.commons.jexl.parser.ASTReference.execute(ASTReference.java:77)
	at org.apache.commons.jexl.parser.ASTReference.value(ASTReference.java:57)
	at org.apache.commons.jexl.parser.ASTElvisNode.value(ASTElvisNode.java:37)
	at org.apache.commons.jexl.parser.ASTExpression.value(ASTExpression.java:54)
	at org.apache.commons.jexl.parser.ASTElvisNode.value(ASTElvisNode.java:40)
	at org.apache.commons.jexl.parser.ASTExpression.value(ASTExpression.java:54)
	at org.apache.commons.jexl.parser.ASTExpressionExpression.value(ASTExpressionExpression.java:56)
	at org.apache.commons.jexl.ExpressionImpl.evaluate(ExpressionImpl.java:80)
	at hudson.ExpressionFactory2$JexlExpression.evaluate(ExpressionFactory2.java:74)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$3.run(CoreTagLibrary.java:134)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.CallTagLibScript.run(CallTagLibScript.java:120)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.CallTagLibScript$1.run(CallTagLibScript.java:99)
	at org.apache.commons.jelly.tags.define.InvokeBodyTag.doTag(InvokeBodyTag.java:91)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.CallTagLibScript$1.run(CallTagLibScript.java:99)
	at org.apache.commons.jelly.tags.define.InvokeBodyTag.doTag(InvokeBodyTag.java:91)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.TagSupport.invokeBody(TagSupport.java:161)
	at org.apache.commons.jelly.tags.core.WhenTag.doTag(WhenTag.java:46)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.TagSupport.invokeBody(TagSupport.java:161)
	at org.apache.commons.jelly.tags.core.ChooseTag.doTag(ChooseTag.java:38)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.CallTagLibScript.run(CallTagLibScript.java:120)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.CallTagLibScript.run(CallTagLibScript.java:120)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.JellyViewScript.run(JellyViewScript.java:95)
	at org.kohsuke.stapler.jelly.IncludeTag.doTag(IncludeTag.java:147)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.CallTagLibScript$1.run(CallTagLibScript.java:99)
	at org.apache.commons.jelly.tags.define.InvokeBodyTag.doTag(InvokeBodyTag.java:91)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.CallTagLibScript.run(CallTagLibScript.java:120)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.JellyViewScript.run(JellyViewScript.java:95)
	at org.kohsuke.stapler.jelly.IncludeTag.doTag(IncludeTag.java:147)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.JellyViewScript.run(JellyViewScript.java:95)
	at org.kohsuke.stapler.jelly.IncludeTag.doTag(IncludeTag.java:147)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.CallTagLibScript$1.run(CallTagLibScript.java:99)
	at org.apache.commons.jelly.tags.define.InvokeBodyTag.doTag(InvokeBodyTag.java:91)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.CallTagLibScript.run(CallTagLibScript.java:120)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.CallTagLibScript$1.run(CallTagLibScript.java:99)
	at org.apache.commons.jelly.tags.define.InvokeBodyTag.doTag(InvokeBodyTag.java:91)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$1.run(CoreTagLibrary.java:98)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.CallTagLibScript.run(CallTagLibScript.java:120)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.CallTagLibScript$1.run(CallTagLibScript.java:99)
	at org.apache.commons.jelly.tags.define.InvokeBodyTag.doTag(InvokeBodyTag.java:91)
	at org.apache.commons.jelly.impl.TagScript.run(TagScript.java:269)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.kohsuke.stapler.jelly.ReallyStaticTagLibrary$1.run(ReallyStaticTagLibrary.java:99)
	at org.apache.commons.jelly.impl.ScriptBlock.run(ScriptBlock.java:95)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.CallTagLibScript.run(CallTagLibScript.java:120)
	at org.apache.commons.jelly.tags.core.CoreTagLibrary$2.run(CoreTagLibrary.java:105)
	at org.kohsuke.stapler.jelly.JellyViewScript.run(JellyViewScript.java:95)
	at org.kohsuke.stapler.jelly.DefaultScriptInvoker.invokeScript(DefaultScriptInvoker.java:63)
	at org.kohsuke.stapler.jelly.DefaultScriptInvoker.invokeScript(DefaultScriptInvoker.java:53)
	at org.kohsuke.stapler.jelly.JellyFacet$1.dispatch(JellyFacet.java:95)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:746)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.MetaClass$5.doDispatch(MetaClass.java:233)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:746)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:649)
	at org.kohsuke.stapler.Stapler.service(Stapler.java:238)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:848)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:686)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1494)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:134)
	at hudson.util.PluginServletFilter.doFilter(PluginServletFilter.java:125)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at hudson.security.csrf.CrumbFilter.doFilter(CrumbFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:84)
	at hudson.security.UnwrapSecurityExceptionFilter.doFilter(UnwrapSecurityExceptionFilter.java:51)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:117)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.providers.anonymous.AnonymousProcessingFilter.doFilter(AnonymousProcessingFilter.java:125)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.rememberme.RememberMeProcessingFilter.doFilter(RememberMeProcessingFilter.java:142)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.AbstractProcessingFilter.doFilter(AbstractProcessingFilter.java:271)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.BasicHeaderProcessor.doFilter(BasicHeaderProcessor.java:93)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.context.HttpSessionContextIntegrationFilter.doFilter(HttpSessionContextIntegrationFilter.java:249)
	at hudson.security.HttpSessionContextIntegrationFilter2.doFilter(HttpSessionContextIntegrationFilter2.java:67)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:76)
	at hudson.security.HudsonFilter.doFilter(HudsonFilter.java:171)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at org.kohsuke.stapler.compression.CompressionFilter.doFilter(CompressionFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:81)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1482)
	at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter(DiagnosticThreadNameFilter.java:30)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1474)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:499)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:137)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:533)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:231)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1086)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:428)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:193)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1020)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:135)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:116)
	at org.eclipse.jetty.server.Server.handle(Server.java:370)
	at org.eclipse.jetty.server.AbstractHttpConnection.handleRequest(AbstractHttpConnection.java:489)
	at org.eclipse.jetty.server.AbstractHttpConnection.headerComplete(AbstractHttpConnection.java:949)
	at org.eclipse.jetty.server.AbstractHttpConnection$RequestHandler.headerComplete(AbstractHttpConnection.java:1011)
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:644)
	at org.eclipse.jetty.http.HttpParser.parseAvailable(HttpParser.java:235)
	at org.eclipse.jetty.server.AsyncHttpConnection.handle(AsyncHttpConnection.java:82)
	at org.eclipse.jetty.io.nio.SelectChannelEndPoint.handle(SelectChannelEndPoint.java:668)
	at org.eclipse.jetty.io.nio.SelectChannelEndPoint$1.run(SelectChannelEndPoint.java:52)
	at winstone.BoundedExecutorService$1.run(BoundedExecutorService.java:77)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.AssertionError: class jenkins.plugins.slack.SlackNotifier$SlackJobProperty is missing its descriptor
	at jenkins.model.Jenkins.getDescriptorOrDie(Jenkins.java:1257)
	at hudson.model.JobProperty.getDescriptor(JobProperty.java:105)
	at hudson.model.JobProperty.getDescriptor(JobProperty.java:79)
	at hudson.model.Descriptor.toMap(Descriptor.java:973)
	at hudson.model.Job.getProperties(Job.java:558)
	... 165 more
```

@OwensCode I only had a handful of pipeline jobs to modify manually and after when the SlackJobProperty section is removed they were ok as well. 
A side effect I noticed - a few other jobs which had injected environment variables into the build and this configuration was removed. I'm guessing this happened when the job was updated with the Environment Injector plugin disabled. 

The stack trace for the pipeline job failure at startup is below. It wasn't very helpful in determining it was a pipeline (or which job for that matter).
```
SEVERE: Failed to initialize Jenkins
hudson.util.HudsonFailedToLoad: java.lang.AssertionError: class jenkins.plugins.slack.SlackNotifier$SlackJobProperty is missing its descriptor
  at hudson.WebAppMain$3.run(WebAppMain.java:234)
Caused by: java.lang.AssertionError: class jenkins.plugins.slack.SlackNotifier$SlackJobProperty is missing its descriptor
  at jenkins.model.Jenkins.getDescriptorOrDie(Jenkins.java:1189)
  at hudson.model.JobProperty.getDescriptor(JobProperty.java:100)
  at hudson.model.JobProperty.getDescriptor(JobProperty.java:75)
  at hudson.model.Descriptor.toMap(Descriptor.java:880)
  at hudson.model.Job.getProperties(Job.java:547)
  at org.jenkinsci.plugins.envinject.migration.EnvInjectMigrationListener.onLoaded(EnvInjectMigrationListener.java:45)
  at jenkins.model.Jenkins.<init>(Jenkins.java:864)
  at hudson.model.Hudson.<init>(Hudson.java:85)
  at hudson.model.Hudson.<init>(Hudson.java:81)
  at hudson.WebAppMain$3.run(WebAppMain.java:225)
```

I wisely created a new Jenkins master with 2.0, rather than install it atop my current Jenkins server (which is happily running jobs in a charming, yet retro manner)  I re-created all of my jobs on the new server. It was harmless because it's not running the Agent servers, so any jobs would just stack up and not run. When I restarted Jenkins 2.0, it was unable to come up due to the Slack error. Once I removed all of the jobs that use the Slack plugin, the server came up normally. I checked and the Slack plugin that was installed is already at the most current release.

hudson.util.HudsonFailedToLoad: java.lang.AssertionError: class jenkins.plugins.slack.SlackNotifier$SlackJobProperty is missing its descriptor
	at hudson.WebAppMain$3.run(WebAppMain.java:234)
Caused by: java.lang.AssertionError: class jenkins.plugins.slack.SlackNotifier$SlackJobProperty is missing its descriptor
	at jenkins.model.Jenkins.getDescriptorOrDie(Jenkins.java:1314)
	at hudson.model.JobProperty.getDescriptor(JobProperty.java:105)
	at hudson.model.JobProperty.getDescriptor(JobProperty.java:79)
	at hudson.model.Descriptor.toMap(Descriptor.java:986)
	at hudson.model.Job.getProperties(Job.java:558)
	at org.jenkinsci.plugins.envinject.migration.EnvInjectMigrationListener.onLoaded(EnvInjectMigrationListener.java:45)
	at jenkins.model.Jenkins.<init>(Jenkins.java:914)
	at hudson.model.Hudson.<init>(Hudson.java:85)
	at hudson.model.Hudson.<init>(Hudson.java:81)
	at hudson.WebAppMain$3.run(WebAppMain.java:225)

I'll continue to tinker with this.

It bothers me that the plugin entirely prevented Jenkins from being accessible, rather than just borking individual jobs. 

Just another datapoint.

tl;dr - set up new J2.0 server. Migrated old jobs to it. Server would not start. Removed jobs that use Slack. Server started.

I have a freshly installed Jenkins 2.7 with the slack plugin version 2.0.1 and I also have this problem:

```
Caused by: java.lang.AssertionError: class jenkins.plugins.slack.SlackNotifier$SlackJobProperty is missing its descriptor
	at jenkins.model.Jenkins.getDescriptorOrDie(Jenkins.java:1301)
	at hudson.model.JobProperty.getDescriptor(JobProperty.java:105)
	at hudson.model.JobProperty.getDescriptor(JobProperty.java:79)
	at hudson.util.DescribableList.get(DescribableList.java:127)
	at hudson.util.DescribableList.rebuild(DescribableList.java:168)
	at hudson.model.Job.doConfigSubmit(Job.java:1219)
	at hudson.model.AbstractProject.doConfigSubmit(AbstractProject.java:795)
	... 74 more
```

Any ideas how to work around it?

I am having the same problem with Jenkins ver 2.7. 

hudson.util.HudsonFailedToLoad: java.lang.AssertionError: class jenkins.plugins.slack.SlackNotifier$SlackJobProperty is missing its descriptor
	at hudson.WebAppMain$3.run(WebAppMain.java:241)
Caused by: java.lang.AssertionError: class jenkins.plugins.slack.SlackNotifier$SlackJobProperty is missing its descriptor
	at jenkins.model.Jenkins.getDescriptorOrDie(Jenkins.java:1301)
	at hudson.model.JobProperty.getDescriptor(JobProperty.java:105)
	at hudson.model.JobProperty.getDescriptor(JobProperty.java:79)
	at hudson.model.Descriptor.toMap(Descriptor.java:982)
	at hudson.model.Job.getProperties(Job.java:559)
	at org.jenkinsci.plugins.envinject.migration.EnvInjectMigrationListener.onLoaded(EnvInjectMigrationListener.java:45)
	at jenkins.model.Jenkins.<init>(Jenkins.java:921)
	at hudson.model.Hudson.<init>(Hudson.java:85)
	at hudson.model.Hudson.<init>(Hudson.java:81)
	at hudson.WebAppMain$3.run(WebAppMain.java:227)

http://imgur.com/sAFAdok

I have tried restarting Jenkins but that hasn't fixed the issue. Any ideas how to work around it?

I am seeing the same problem in Jenkins V1.609.18.3. I see the stack trace when trying to save configuration changes on some (but not all) jobs, and the changes are not saved.

javax.servlet.ServletException: java.lang.AssertionError: class jenkins.plugins.slack.SlackNotifier$SlackJobProperty is missing its descriptor

I am disabling the plugin until there is a fix, as I can't have an optional plugin breaking some of the basic functions of Jenkins.

--- a/src/main/java/jenkins/plugins/slack/SlackNotifier.java
+++ b/src/main/java/jenkins/plugins/slack/SlackNotifier.java
@@ -6,7 +6,9 @@ import hudson.Launcher;
 import hudson.model.AbstractBuild;
 import hudson.model.AbstractProject;
 import hudson.model.BuildListener;
+import hudson.model.Item;
 import hudson.model.Descriptor;
+import hudson.model.Job;
 import hudson.model.listeners.ItemListener;
 import hudson.tasks.BuildStepDescriptor;
 import hudson.tasks.BuildStepMonitor;
@@ -15,14 +17,21 @@ import hudson.tasks.Publisher;
 import hudson.util.FormValidation;
 import jenkins.model.Jenkins;
 import jenkins.model.JenkinsLocationConfiguration;
+import jenkins.plugins.slack.config.AbstractProjectConfigMigrator;
+import jenkins.plugins.slack.config.ItemConfigMigrator;
+import jenkins.plugins.slack.config.JobConfigMigrator;
 import net.sf.json.JSONObject;
+
 import org.apache.commons.lang.StringUtils;
 import org.kohsuke.stapler.DataBoundConstructor;
 import org.kohsuke.stapler.QueryParameter;
 import org.kohsuke.stapler.StaplerRequest;
 import org.kohsuke.stapler.export.Exported;
+import org.springframework.util.ReflectionUtils;
 
 import java.io.IOException;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
 import java.util.Map;
 import java.util.logging.Level;
 import java.util.logging.Logger;
@@ -58,12 +67,24 @@ public class SlackNotifier extends Notifier {
         return teamDomain;
     }
 
+    public void setTeamDomain(final String teamDomain) {
+        this.teamDomain = teamDomain;
+    }
+
     public String getRoom() {
         return room;
     }
 
+    public void setRoom(String room) {
+        this.room = room;
+    }
+
     public String getAuthToken() {
         return authToken;
+    }
+
+    public void setAuthToken(String authToken) {
+        this.authToken = authToken;
     }
 
     public String getBuildServerUrl() {
@@ -126,6 +147,54 @@ public class SlackNotifier extends Notifier {
 
     public String getCustomMessage() {
         return customMessage;
+    }
+
+    public void setStartNotification(boolean startNotification) {
+        this.startNotification = startNotification;
+    }
+
+    public void setNotifySuccess(boolean notifySuccess) {
+        this.notifySuccess = notifySuccess;
+    }
+
+    public void setCommitInfoChoice(CommitInfoChoice commitInfoChoice) {
+        this.commitInfoChoice = commitInfoChoice;
+    }
+
+    public void setNotifyAborted(boolean notifyAborted) {
+        this.notifyAborted = notifyAborted;
+    }
+
+    public void setNotifyFailure(boolean notifyFailure) {
+        this.notifyFailure = notifyFailure;
+    }
+
+    public void setNotifyNotBuilt(boolean notifyNotBuilt) {
+        this.notifyNotBuilt = notifyNotBuilt;
+    }
+
+    public void setNotifyUnstable(boolean notifyUnstable) {
+        this.notifyUnstable = notifyUnstable;
+    }
+
+    public void setNotifyBackToNormal(boolean notifyBackToNormal) {
+        this.notifyBackToNormal = notifyBackToNormal;
+    }
+
+    public void setIncludeTestSummary(boolean includeTestSummary) {
+        this.includeTestSummary = includeTestSummary;
+    }
+
+    public void setNotifyRepeatedFailure(boolean notifyRepeatedFailure) {
+        this.notifyRepeatedFailure = notifyRepeatedFailure;
+    }
+
+    public void setIncludeCustomMessage(boolean includeCustomMessage) {
+        this.includeCustomMessage = includeCustomMessage;
+    }
+
+    public void setCustomMessage(String customMessage) {
+        this.customMessage = customMessage;
     }
 
     @DataBoundConstructor
@@ -465,66 +534,21 @@ public class SlackNotifier extends Notifier {
     }
 
     @Extension public static final class Migrator extends ItemListener {
-        @SuppressWarnings("deprecation")
         @Override
         public void onLoaded() {
             logger.info("Starting Settings Migration Process");
-            for (AbstractProject<?, ?> p : Jenkins.getInstance().getAllItems(AbstractProject.class)) {
-                logger.info("processing Job: " + p.getName());
-
-                final SlackJobProperty slackJobProperty = p.getProperty(SlackJobProperty.class);
-
-                if (slackJobProperty == null) {
-                    logger.info(String
-                            .format("Configuration is already up to date for \"%s\", skipping migration",
-                                    p.getName()));
+            
+            ItemConfigMigrator migrator = new ItemConfigMigrator();
+            
+            for (Item item : Jenkins.getInstance().getAllItems()) {
+                if (!migrator.migrate(item)) {
+                    logger.info(String.format("Skipping job \"%s\" with type %s", item.getName(),
+                            item.getClass().getName()));
                     continue;
                 }
-
-                SlackNotifier slackNotifier = p.getPublishersList().get(SlackNotifier.class);
-                
-                if (slackNotifier == null) {
-                    logger.info(String
-                            .format("Configuration does not have a notifier for \"%s\", not migrating settings",
-                                    p.getName()));
-                } else {
-
-                    //map settings
-                    if (StringUtils.isBlank(slackNotifier.teamDomain)) {
-                        slackNotifier.teamDomain = slackJobProperty.getTeamDomain();
-                    }
-                    if (StringUtils.isBlank(slackNotifier.authToken)) {
-                        slackNotifier.authToken = slackJobProperty.getToken();
-                    }
-                    if (StringUtils.isBlank(slackNotifier.room)) {
-                        slackNotifier.room = slackJobProperty.getRoom();
-                    }
-                    
-                    slackNotifier.startNotification = slackJobProperty.getStartNotification();
-    
-                    slackNotifier.notifyAborted = slackJobProperty.getNotifyAborted();
-                    slackNotifier.notifyFailure = slackJobProperty.getNotifyFailure();
-                    slackNotifier.notifyNotBuilt = slackJobProperty.getNotifyNotBuilt();
-                    slackNotifier.notifySuccess = slackJobProperty.getNotifySuccess();
-                    slackNotifier.notifyUnstable = slackJobProperty.getNotifyUnstable();
-                    slackNotifier.notifyBackToNormal = slackJobProperty.getNotifyBackToNormal();
-                    slackNotifier.notifyRepeatedFailure = slackJobProperty.getNotifyRepeatedFailure();
-    
-                    slackNotifier.includeTestSummary = slackJobProperty.includeTestSummary();
-                    slackNotifier.commitInfoChoice = slackJobProperty.getShowCommitList() ? CommitInfoChoice.AUTHORS_AND_TITLES : CommitInfoChoice.NONE;
-                    slackNotifier.includeCustomMessage = slackJobProperty.includeCustomMessage();
-                    slackNotifier.customMessage = slackJobProperty.getCustomMessage();
-                }
-
-                try {
-                    //property section is not used anymore - remove
-                    p.removeProperty(SlackJobProperty.class);
-                    p.save();
-                    logger.info("Configuration updated successfully");
-                } catch (IOException e) {
-                    logger.log(Level.SEVERE, e.getMessage(), e);
-                }
             }
+
+            logger.info("Completed Settings Migration Process");
         }
     }
 }

--- a/src/main/java/jenkins/plugins/slack/config/AbstractProjectConfigMigrator.java
+++ b/src/main/java/jenkins/plugins/slack/config/AbstractProjectConfigMigrator.java
@@ -1,0 +1,98 @@
+package jenkins.plugins.slack.config;
+
+import hudson.model.AbstractProject;
+
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import jenkins.plugins.slack.CommitInfoChoice;
+import jenkins.plugins.slack.SlackNotifier;
+import jenkins.plugins.slack.SlackNotifier.SlackJobProperty;
+
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * Configuration migrator for migrating the Slack plugin configuration for a
+ * {@link AbstractProject} from the 1.8 format to the 2.0 format. It does so by
+ * removing the SlackJobProperty from the job properties (if there is one) and
+ * moving the Slack notification settings to a {@link SlackNotifier} in the list
+ * of publishers (if there is one).
+ */
+@SuppressWarnings("deprecation")
+public class AbstractProjectConfigMigrator {
+
+    private static final Logger logger = Logger.getLogger(AbstractProjectConfigMigrator.class
+            .getName());
+
+    public void migrate(final AbstractProject<?, ?> project) {
+
+        logger.info(String.format("Migrating project \"%s\" with type %s", project.getName(),
+                project.getClass().getName()));
+
+        final SlackJobProperty slackJobProperty = project.getProperty(SlackJobProperty.class);
+
+        if (slackJobProperty == null) {
+            logger.info(String.format(
+                    "Configuration is already up to date for \"%s\", skipping migration",
+                    project.getName()));
+            return;
+        }
+
+        SlackNotifier slackNotifier = project.getPublishersList().get(SlackNotifier.class);
+
+        if (slackNotifier == null) {
+            logger.info(String.format(
+                    "Configuration does not have a notifier for \"%s\", not migrating settings",
+                    project.getName()));
+        } else {
+            updateSlackNotifier(slackNotifier, slackJobProperty);
+        }
+
+        try {
+            // property section is not used anymore - remove
+            project.removeProperty(SlackJobProperty.class);
+            project.save();
+            logger.info("Configuration updated successfully");
+        } catch (IOException e) {
+            logger.log(Level.SEVERE, e.getMessage(), e);
+        }
+    }
+
+    private void updateSlackNotifier(final SlackNotifier slackNotifier,
+            final SlackJobProperty slackJobProperty) {
+
+        if (StringUtils.isBlank(slackNotifier.getTeamDomain())) {
+            slackNotifier.setTeamDomain(slackJobProperty.getTeamDomain());
+        }
+        if (StringUtils.isBlank(slackNotifier.getAuthToken())) {
+            slackNotifier.setAuthToken(slackJobProperty.getToken());
+        }
+        if (StringUtils.isBlank(slackNotifier.getRoom())) {
+            slackNotifier.setRoom(slackJobProperty.getRoom());
+        }
+
+        slackNotifier.setStartNotification(slackJobProperty.getStartNotification());
+
+        slackNotifier.setNotifyAborted(slackJobProperty.getNotifyAborted());
+        slackNotifier.setNotifyFailure(slackJobProperty.getNotifyFailure());
+        slackNotifier.setNotifyNotBuilt(slackJobProperty.getNotifyNotBuilt());
+        slackNotifier.setNotifySuccess(slackJobProperty.getNotifySuccess());
+        slackNotifier.setNotifyUnstable(slackJobProperty.getNotifyUnstable());
+        slackNotifier.setNotifyBackToNormal(slackJobProperty.getNotifyBackToNormal());
+        slackNotifier.setNotifyRepeatedFailure(slackJobProperty.getNotifyRepeatedFailure());
+
+        slackNotifier.setIncludeTestSummary(slackJobProperty.includeTestSummary());
+        slackNotifier.setCommitInfoChoice(getCommitInfoChoice(slackJobProperty));
+        slackNotifier.setIncludeCustomMessage(slackJobProperty.includeCustomMessage());
+        slackNotifier.setCustomMessage(slackJobProperty.getCustomMessage());
+    }
+
+    private CommitInfoChoice getCommitInfoChoice(final SlackJobProperty slackJobProperty) {
+        if (slackJobProperty.getShowCommitList()) {
+            return CommitInfoChoice.AUTHORS_AND_TITLES;
+        } else {
+            return CommitInfoChoice.NONE;
+        }
+    }
+}

--- a/src/main/java/jenkins/plugins/slack/config/ItemConfigMigrator.java
+++ b/src/main/java/jenkins/plugins/slack/config/ItemConfigMigrator.java
@@ -1,0 +1,112 @@
+package jenkins.plugins.slack.config;
+
+import java.lang.reflect.Method;
+
+import org.springframework.util.ReflectionUtils;
+
+import hudson.model.AbstractProject;
+import hudson.model.Item;
+import hudson.model.Job;
+import hudson.util.DescribableList;
+
+/**
+ * Configuration migrator for migrating the Slack plugin configuration from the
+ * 1.8 format to the 2.0 format, for an item like a job or project. Mainly just
+ * decides what needs to be done and delegates the actual migration work to
+ * other migrators.
+ * 
+ * @see AbstractProjectConfigMigrator
+ * @see JobConfigMigrator
+ * @see ItemWithTemplateConfigMigrator
+ */
+public class ItemConfigMigrator {
+
+    private final AbstractProjectConfigMigrator projectMigrator;
+    private final JobConfigMigrator jobMigrator;
+    private final ItemWithTemplateConfigMigrator templateMigrator;
+
+    public ItemConfigMigrator() {
+        projectMigrator = new AbstractProjectConfigMigrator();
+        jobMigrator = new JobConfigMigrator();
+        templateMigrator = new ItemWithTemplateConfigMigrator(projectMigrator);
+    }
+    
+    /**
+     * Constructor for injecting migrators for testing.
+     */
+    protected ItemConfigMigrator(AbstractProjectConfigMigrator projectMigrator,
+            JobConfigMigrator jobMigrator, ItemWithTemplateConfigMigrator templateMigrator) {
+        this.projectMigrator = projectMigrator;
+        this.jobMigrator = jobMigrator;
+        this.templateMigrator = templateMigrator;
+    }
+
+    /**
+     * Migrate configuration for a {@link Item} from the 1.8 format to the 2.0
+     * format. This primarily removes job properties and adds them to a
+     * notifier.
+     * 
+     * @param item
+     *            Item to migrate
+     * @return true if migration was attempted for an expected scenario; false
+     *         if migration was not attempted due to an unexpected data type or
+     *         other reason
+     */
+    public boolean migrate(Item item) {
+
+        // Attempt migrations in priority order
+        return migrateAbstractProject(item) || migrateJobWithoutPublishersList(item)
+                || templateMigrator.migrate(item);
+    }
+
+    /**
+     * Migrate an item if it is a subclass of AbstractProject.
+     * 
+     * @param item
+     *            Item to migrate
+     * @return true if migration was attempted
+     */
+    private boolean migrateAbstractProject(Item item) {
+
+        if (item instanceof AbstractProject) {
+            AbstractProject<?, ?> project = (AbstractProject<?, ?>) item;
+            projectMigrator.migrate(project);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Migrate an item if it is subclass of Job. It might have Slack job
+     * properties that need to be removed in the migration, but does not handle
+     * the possibility of a Job that has publishers. That should theoretically
+     * not happen because then it should be a subclass of AbstractProject, but
+     * we want to avoid losing settings. So if it looks like it has publishers,
+     * migration is skipped.
+     * 
+     * @param item
+     *            Item to migrate
+     * @return true if migration was attempted
+     */
+    private boolean migrateJobWithoutPublishersList(Item item) {
+        if (item instanceof Job) {
+            if (!hasMethodGetPublishersList(item)) {
+                Job<?, ?> job = (Job<?, ?>) item;
+                jobMigrator.migrate(job);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean hasMethodGetPublishersList(Item item) {
+
+        Method method = ReflectionUtils.findMethod(item.getClass(), "getPublishersList");
+
+        if (method != null && method.getReturnType().equals(DescribableList.class)) {
+            return true;
+        }
+
+        return false;
+    }
+}

--- a/src/main/java/jenkins/plugins/slack/config/ItemWithTemplateConfigMigrator.java
+++ b/src/main/java/jenkins/plugins/slack/config/ItemWithTemplateConfigMigrator.java
@@ -1,0 +1,100 @@
+package jenkins.plugins.slack.config;
+
+import java.lang.reflect.Method;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.springframework.util.ReflectionUtils;
+
+import hudson.model.AbstractProject;
+import hudson.model.Item;
+import hudson.model.Job;
+
+/**
+ * Configuration migrator for migrating the Slack plugin configuration for an
+ * {@link AbstractProject} that belongs to an {@link Item} as a template, from
+ * the 1.8 format to the 2.0 format.
+ * 
+ * <p>
+ * This is a workaround for installations that use the multi-branch-project
+ * plugin, which uses a template to configure jobs for all branches in a repo.
+ * The template will be updated by an {@link AbstractProjectConfigMigrator}.
+ * </p>
+ */
+public class ItemWithTemplateConfigMigrator {
+
+    private static final Logger logger = Logger.getLogger(ItemWithTemplateConfigMigrator.class
+            .getName());
+
+    private AbstractProjectConfigMigrator projectMigrator;
+
+    /**
+     * Default constructor.
+     * 
+     * @param projectMigrator
+     *            Migrator to be used for migrating the template
+     */
+    public ItemWithTemplateConfigMigrator(final AbstractProjectConfigMigrator projectMigrator) {
+        this.projectMigrator = projectMigrator;
+    }
+
+    /**
+     * Migrate an item if it has a template that is a subclass of
+     * AbstractProject.
+     * 
+     * @param item
+     *            Item to migrate
+     * @return true if migration was attempted on a template
+     */
+    public boolean migrate(final Item item) {
+        AbstractProject<?, ?> project = getTemplateProject(item);
+
+        if (project != null) {
+            projectMigrator.migrate(project);
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Examine an Item to determine if it has a "getTemplate" method that
+     * returns an AbstractProject, and return the AbstractProject if it does.
+     * 
+     * @param item
+     *            Item to examine
+     * @return AbstractProject that is returned by item.getTemplate(), or null
+     */
+    private AbstractProject<?, ?> getTemplateProject(final Item item) {
+
+        logger.log(Level.FINE,
+                String.format("Checking \"%s\" for AbstractProject template", item.getName()));
+
+        Method getTemplate = ReflectionUtils.findMethod(item.getClass(), "getTemplate");
+
+        if (getTemplate == null) {
+            logger.log(Level.FINE, "No template getter method found");
+            return null;
+        }
+
+        Object obj = null;
+
+        try {
+            obj = getTemplate.invoke(item);
+        } catch (Exception e) {
+            logger.info("Error getting \"template\" value: " + e.getMessage());
+            return null;
+        }
+
+        if (obj == null) {
+            logger.log(Level.FINE, "Item has no template");
+        } else if (obj instanceof AbstractProject) {
+            return (AbstractProject<?, ?>) obj;
+        } else {
+            logger.log(Level.FINE, "Template is not an AbstractProject; type is: "
+                    + obj.getClass().getName());
+        }
+
+        return null;
+    }
+}

--- a/src/main/java/jenkins/plugins/slack/config/JobConfigMigrator.java
+++ b/src/main/java/jenkins/plugins/slack/config/JobConfigMigrator.java
@@ -1,0 +1,51 @@
+package jenkins.plugins.slack.config;
+
+import hudson.model.Job;
+
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import jenkins.plugins.slack.SlackNotifier.SlackJobProperty;
+
+/**
+ * Configuration migrator for migrating the Slack plugin configuration for a
+ * {@link Job} from the 1.8 format to the 2.0 format. It does so by removing the
+ * SlackJobProperty from the job properties (if there is one).
+ * 
+ * <p>
+ * SlackJobProperty settings are usually migrated to a publisher, but there are
+ * no publishers in a Job so the settings are lost. For this reason, <strong>be
+ * careful of how you use this migrator.</strong>.
+ * </p>
+ */
+@SuppressWarnings("deprecation")
+public class JobConfigMigrator {
+
+    private static final Logger logger = Logger.getLogger(JobConfigMigrator.class.getName());
+
+    public void migrate(final Job<?, ?> job) {
+
+        logger.info(String.format("Migrating job \"%s\" with type %s", job.getName(), job
+                .getClass().getName()));
+
+        final SlackJobProperty slackJobProperty = job.getProperty(SlackJobProperty.class);
+
+        if (slackJobProperty == null) {
+            logger.info(String.format(
+                    "Configuration is already up to date for \"%s\", skipping migration",
+                    job.getName()));
+            return;
+        }
+
+        try {
+            // property section is not used anymore - remove
+            job.removeProperty(SlackJobProperty.class);
+            job.save();
+            logger.info(String.format("Configuration for \"%s\" updated successfully",
+                    job.getName()));
+        } catch (IOException e) {
+            logger.log(Level.SEVERE, e.getMessage(), e);
+        }
+    }
+}

--- a/src/test/java/jenkins/plugins/slack/config/ItemConfigMigratorTest.java
+++ b/src/test/java/jenkins/plugins/slack/config/ItemConfigMigratorTest.java
@@ -1,0 +1,111 @@
+package jenkins.plugins.slack.config;
+
+import jenkins.model.AbstractTopLevelItem;
+import jenkins.model.Jenkins;
+import hudson.model.Descriptor;
+import hudson.model.Item;
+import hudson.model.ItemGroup;
+import hudson.model.ViewJob;
+import hudson.model.AbstractProject;
+import hudson.model.Job;
+import hudson.tasks.Publisher;
+import hudson.util.DescribableList;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+
+import static org.junit.Assert.*;
+import static org.mockito.Matchers.*;
+import static org.mockito.Mockito.*;
+
+public class ItemConfigMigratorTest {
+    
+    private AbstractProjectConfigMigrator projectMigrator;
+    private JobConfigMigrator jobMigrator;
+    private ItemWithTemplateConfigMigrator templateMigrator;
+    private ItemConfigMigrator migrator;
+    
+    private Jenkins jenkins;
+    
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+    
+    @Before
+    public void setUp() {
+        jenkins = j.getInstance();
+        
+        projectMigrator = mock(AbstractProjectConfigMigrator.class);
+        jobMigrator = mock(JobConfigMigrator.class);
+        templateMigrator = mock(ItemWithTemplateConfigMigrator.class);
+        
+        migrator = new ItemConfigMigrator(projectMigrator, jobMigrator, templateMigrator);
+    }
+
+    @Test
+    public void testMigrate_AbstractProject() {
+        AbstractProject<?, ?> item = mock(AbstractProject.class);
+        
+        assertTrue(migrator.migrate(item));
+        
+        verify(projectMigrator).migrate(any(AbstractProject.class));
+        verify(jobMigrator, never()).migrate(any(Job.class));
+        verify(templateMigrator, never()).migrate(any(AbstractProject.class));
+    }
+    
+    @Test
+    public void testMigrate_Job() {
+        Job<?, ?> item = mock(Job.class);
+        
+        assertTrue(migrator.migrate(item));
+        
+        verify(projectMigrator, never()).migrate(any(AbstractProject.class));
+        verify(jobMigrator).migrate(any(Job.class));
+        verify(templateMigrator, never()).migrate(any(AbstractProject.class));
+    }
+    
+    @Test
+    public void testMigrate_JobWithPublishersList() {
+        Job<?, ?> item = new JobWithPublishers(jenkins, "Random Name");
+        
+        doReturn(false).when(templateMigrator).migrate(item);
+        
+        assertFalse(migrator.migrate(item));
+        
+        verify(projectMigrator, never()).migrate(any(AbstractProject.class));
+        verify(jobMigrator, never()).migrate(any(Job.class));
+        verify(templateMigrator).migrate(any(AbstractProject.class));
+    }
+    
+    @Test
+    public void testMigrate_ItemWithTemplate() {
+        
+        Item item = mock(Item.class);
+        
+        doReturn(true).when(templateMigrator).migrate(item);
+        
+        assertTrue(migrator.migrate(item));
+        
+        verify(projectMigrator, never()).migrate(any(AbstractProject.class));
+        verify(jobMigrator, never()).migrate(any(Job.class));
+        verify(templateMigrator).migrate(eq(item));
+    }
+    
+    @SuppressWarnings("rawtypes")
+    private static class JobWithPublishers extends ViewJob {
+        
+        public JobWithPublishers(ItemGroup parent, String name) {
+            super(parent, name);
+        }
+
+        @Override
+        public void reload() {
+        }
+        
+        @SuppressWarnings("unused")
+        public DescribableList<Publisher,Descriptor<Publisher>> getPublishersList() {
+            return null;
+        }
+    }
+}

--- a/src/test/java/jenkins/plugins/slack/config/ItemWithTemplateConfigMigratorTest.java
+++ b/src/test/java/jenkins/plugins/slack/config/ItemWithTemplateConfigMigratorTest.java
@@ -1,0 +1,95 @@
+package jenkins.plugins.slack.config;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.*;
+import hudson.model.ItemGroup;
+import hudson.model.AbstractProject;
+import hudson.model.FreeStyleProject;
+import jenkins.model.AbstractTopLevelItem;
+import jenkins.model.Jenkins;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+
+@SuppressWarnings("deprecation")
+public class ItemWithTemplateConfigMigratorTest {
+
+    private Jenkins jenkins;
+    private ItemWithTemplateConfigMigrator migrator;
+    private AbstractProjectConfigMigrator projectMigratorMock;
+
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+
+    @Before
+    public void setUp() {
+        jenkins = j.getInstance();
+        projectMigratorMock = mock(AbstractProjectConfigMigrator.class);
+        migrator = new ItemWithTemplateConfigMigrator(projectMigratorMock);
+    }
+
+    @Test
+    public void testMigrate_WithNoTemplateMethod() {
+        FreeStyleProject project = new FreeStyleProject(jenkins, "Test_Slack_Plugin");
+        assertFalse("Should not be able to find template project", migrator.migrate(project));
+        verify(projectMigratorMock, never()).migrate(any(AbstractProject.class));
+    }
+
+    @Test
+    public void testMigrate_WithNoTemplate() {
+        ProjectWithTemplate project = new ProjectWithTemplate(jenkins, "Test", null);
+        assertFalse("Should be no template project", migrator.migrate(project));
+        verify(projectMigratorMock, never()).migrate(any(AbstractProject.class));
+    }
+
+    @Test
+    public void testMigrate_WithTemplateWrongType() {
+        ProjectWithTemplate project = new ProjectWithTemplate(jenkins, "Test", "NOT A PROJECT");
+        assertFalse("Template is not correct type", migrator.migrate(project));
+        verify(projectMigratorMock, never()).migrate(any(AbstractProject.class));
+    }
+
+    @Test
+    public void testMigrate_WithTemplate() {
+        FreeStyleProject template = new FreeStyleProject(jenkins, "Test_Slack_Plugin");
+        ProjectWithTemplate project = new ProjectWithTemplate(jenkins, "Test", template);
+        assertTrue("Template was migrated", migrator.migrate(project));
+        verify(projectMigratorMock, times(1)).migrate(eq(template));
+    }
+
+    @Test
+    public void testMigrate_WithException() {
+        FreeStyleProject template = new FreeStyleProject(jenkins, "Test_Slack_Plugin");
+        ProjectWithTemplate project = new ProjectWithTemplate(jenkins, "Test", template, true);
+        assertFalse(migrator.migrate(project));
+        verify(projectMigratorMock, never()).migrate(any(AbstractProject.class));
+    }
+
+    private static class ProjectWithTemplate extends AbstractTopLevelItem {
+        private Object template;
+        private boolean doThrow = false;
+
+        public ProjectWithTemplate(ItemGroup<?> group, String name, Object template) {
+            super(group, name);
+            this.template = template;
+        }
+
+        public ProjectWithTemplate(ItemGroup<?> group, String name, Object template, boolean doThrow) {
+            super(group, name);
+            this.template = template;
+            this.doThrow = doThrow;
+        }
+
+        @SuppressWarnings("unused")
+        public Object getTemplate() {
+            if (doThrow) {
+                throw new RuntimeException("Something bad happened");
+            }
+            return template;
+        }
+    }
+}

--- a/src/test/java/jenkins/plugins/slack/config/JobConfigMigratorTest.java
+++ b/src/test/java/jenkins/plugins/slack/config/JobConfigMigratorTest.java
@@ -1,0 +1,56 @@
+package jenkins.plugins.slack.config;
+
+import static org.mockito.Matchers.*;
+import static org.mockito.Mockito.*;
+import hudson.model.Job;
+
+import java.io.IOException;
+
+import jenkins.plugins.slack.SlackNotifier.SlackJobProperty;
+
+import org.junit.Before;
+import org.junit.Test;
+
+@SuppressWarnings("deprecation")
+public class JobConfigMigratorTest {
+    
+    private JobConfigMigrator migrator;
+    
+    @Before
+    public void setUp() {
+        migrator = new JobConfigMigrator();
+    }
+
+    @Test
+    public void testMigrate_WithNoSlackJobProperty() throws Exception {
+        Job<?, ?> project = mock(Job.class);
+        migrator.migrate(project);
+        verify(project).getProperty(eq(SlackJobProperty.class));
+        verify(project, never()).removeProperty(eq(SlackJobProperty.class));
+    }
+    
+    @Test
+    public void testMigrate_WithSlackJobProperty() throws Exception {
+        SlackJobProperty jobProperty = mock(SlackJobProperty.class);
+        
+        Job<?, ?> project = mock(Job.class);
+        when(project.getProperty(eq(SlackJobProperty.class))).thenReturn(jobProperty);
+        
+        migrator.migrate(project);
+        verify(project).getProperty(eq(SlackJobProperty.class));
+        verify(project).removeProperty(eq(SlackJobProperty.class));
+    }
+    
+    @Test
+    public void testMigrate_WithIOExceptionOnSave() throws Exception {
+        SlackJobProperty jobProperty = mock(SlackJobProperty.class);
+        
+        Job<?, ?> project = mock(Job.class);
+        when(project.getProperty(eq(SlackJobProperty.class))).thenReturn(jobProperty);
+        doThrow(new IOException("Something bad happened")).when(project).save();
+        
+        migrator.migrate(project);
+        
+        // exception caught and nothing bad happened
+    }
+}

